umapi_client/functional.py

```diff
@@ -69,21 +69,35 @@ class UserAction(Action):
     _domain_regex = re.compile(r"^" + _dns_pattern + r"$", re.UNICODE)
 
     @classmethod
-    def _validate(cls, email=None, username=None, domain=None):
+    def _validate_email(cls, email=None):
         """
-        Validates the specified user attributes against their specifications.
+        Validates user email (or username if it's assumed to be in email format)
         Input values must be strings (standard or unicode).  Throws ArgumentError if any input is invalid
         :param email: an email address
-        :param username: a username
-        :param domain: a domain
         """
         if email and not cls._email_regex.match(email):
             raise ArgumentError("'%s': Illegal email format (must be ascii, unquoted, with no comment part)" % (email,))
-        if domain and not cls._domain_regex.match(domain):
-            raise ArgumentError("'%s': Illegal domain format" % (domain,))
+
+    @classmethod
+    def _validate_username(cls, username=None):
+        """
+        Validates the specified username (assuming it should be non-email)
```

Comment:
This is not correct anymore, because we allow email-formatted usernames that don't match the user's email in Federated.  Whether or not the username is in email format is determined by the SAML settings on the directory, which we don't have access to.  I think you need to be permissive here and allow emails and non-emails both as usernames.

Reply:
The validation conditions in the init function haven't changed.  When creating the outer UserAction object, the username must match the email address.  If they don't match, then validate the username as a non-email username.  That is why we need separate validator functions for username and email address.  It also allows us to validate the username as an email address in the update function.

Reply:
That's overly restrictive, because it has never been the case that we require email-style usernames to match user emails in federated.  See comments below for more color about this.  It's true that UMAPI always required this, but that was because UMAPI was made in the single-domain directory days so the email domain had to match and it assumed that people wouldn't bother using different usernames from the user email.  But the admin console always allowed this, even in the single-domain directory days.

Reply:
I don't understand how console behavior is relevant to this issue.  Regardless of what the console allows or doesn't allow, the UMAPI doesn't allow users to be created with different email-type usernames and email addresses.

If I make a call like this one (where the domains are different), I get an error.

```json
[
  {"user": "test.unemail.0001@adobeccetest.com", "requestID": "action_1", "do": [{"createFederatedID": {"email": "test.unemail.0001@sf.adobeccetest.com", "option": "ignoreIfAlreadyExists", "firstname": "Test", "lastname": "User Email 0001", "country": "US"}}
  ]}
]
```

```json
{
    "completed": 0,
    "notCompleted": 1,
    "completedInTestMode": 0,
    "result": "error",
    "errors": [
        {
            "index": 0,
            "step": 0,
            "message": "User name must match email address",
            "errorCode": "error.user.must_match_email",
            "requestID": "action_1",
            "user": "test.unemail.0001@adobeccetest.com"
        }
    ]
}
```

If we want to loosen validation and allow the client to do things that the UMAPI doesn't allow, then we may as well remove validation altogether and just rely on the UMAPI to accept or reject creates and updates.  I'm OK with doing that, but I can also see the benefits of preemptive validation.  If we decide to keep the validation functions, their behavior ought to be consistent with UMAPI behavior.